skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,433 @@
+
+        Args:
+            X (FDataBasis):
+                the functional data object to be analysed in basis
+                representation
+            y (None, not used):
+                only present for convention of a fit function
+
+        Returns:
+            self (object)
+
+        References:
+            .. [RS05-8-4-2] Ramsay, J., Silverman, B. W. (2005). Basis function
+                expansion of the functions. In *Functional Data Analysis*
+                (pp. 161-164). Springer.
+
+        """
+
+        # the maximum number of components is established by the target basis
+        # if the target basis is available.
+        n_basis = self.components_basis.n_basis if self.components_basis \
+            else X.basis.n_basis
+        n_samples = X.n_samples
+
+        # check that the number of components is smaller than the sample size
+        if self.n_components > X.n_samples:
+            raise AttributeError("The sample size must be bigger than the "
+                                 "number of components")
+
+        # check that we do not exceed limits for n_components as it should
+        # be smaller than the number of attributes of the basis
+        if self.n_components > n_basis:
+            raise AttributeError("The number of components should be "
+                                 "smaller than the number of attributes of "
+                                 "target principal components' basis.")
+
+        # if centering is True then subtract the mean function to each function
+        # in FDataBasis
+        if self.centering:
+            meanfd = X.mean()
+            # consider moving these lines to FDataBasis as a centering function
+            # subtract from each row the mean coefficient matrix
+            X.coefficients -= meanfd.coefficients
+
+        # setup principal component basis if not given
+        if self.components_basis:
+            # First fix domain range if not already done
+            self.components_basis.domain_range = X.basis.domain_range
+            g_matrix = self.components_basis.gram_matrix()
+            # the matrix that are in charge of changing the computed principal
+            # components to target matrix is essentially the inner product
+            # of both basis.
+            j_matrix = X.basis.inner_product(self.components_basis)
+        else:
+            # if no other basis is specified we use the same basis as the passed
+            # FDataBasis Object
+            self.components_basis = X.basis.copy()
+            g_matrix = self.components_basis.gram_matrix()
+            j_matrix = g_matrix
+
+        # make g matrix symmetric, referring to Ramsay's implementation
+        g_matrix = (g_matrix + np.transpose(g_matrix)) / 2
+
+        # Apply regularization / penalty if applicable
+        if self.regularization_parameter > 0:
+            # obtain regularization matrix
+            regularization_matrix = self.components_basis.penalty(
```

Comment:
This call is going to change with #189 .

Reply:
I noticed that and it is on my todo list, just waiting for your pull request to merge and then I will change it to a single parameter.